File: palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> st;
        ListNode * p=head;
        while(p!=nullptr)
        {
            st.push(p->val);
            p=p->next;
        }

        p=head;
        while(!st.empty())
        {
            int x=st.top();
            if(x!=p->val)
                return false;
            st.pop();
            p=p->next;
        }
        return true;

    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.py
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        
        fast = slow = head
        rev = None

        while fast and fast.next:
            fast = fast.next.next
            temp = rev
            rev = slow
            slow = slow.next
            rev.next = temp

        if fast:
            slow = slow.next

        while rev:
            if rev.val != slow.val:
                return False

            slow = slow.next
            rev = rev.next
        
        return True
